include/souper/Infer/Interpreter.h

```diff
@@ -0,0 +1,67 @@
+#ifndef SOUPER_INTERPRTER_H
+#define SOUPER_INTERPRTER_H
+
+#include "llvm/Support/KnownBits.h"
+#include "llvm/IR/ConstantRange.h"
+
+#include "souper/Inst/Inst.h"
+
+namespace souper {
+
+struct EvalValue {
+  enum class Kind {Val, Poison, Undef, Unimplemented};
+
+  EvalValue() : K(Kind::Unimplemented) {}
+  EvalValue(llvm::APInt Val) : Value{Val}, K(Kind::Val) {};
+
+  llvm::Optional<llvm::APInt> Value;
```

Comment:
I don't feel too strongly about it, but maybe you want to just have an APInt here and we don't care about its value if there UB or whatever?